engines/cegar_ops_uf.cpp

```diff
@@ -0,0 +1,261 @@
+**   Ahmed Irfan
+** This file is part of the pono project.
+** Copyright (c) 2019 by the authors listed in the file AUTHORS
+** in the top-level source directory) and their institutional affiliations.
+** All rights reserved.  See the file LICENSE in the top-level source
+** directory for licensing information.\endverbatim
+**
+** \brief A simple CEGAR loop that abstracts operators with uninterpreted
+**        functions and refines by concretizing the operators
+**
+**/
+
+#include "engines/cegar_ops_uf.h"
+
+#include "core/fts.h"
+#include "core/rts.h"
+
+#include "engines/ic3ia.h"
+
+#include "smt/available_solvers.h"
+
+#include "utils/exceptions.h"
+#include "utils/logger.h"
+#include "utils/make_provers.h"
+#include "utils/ts_manipulation.h"
+#include "utils/term_analysis.h"
+
+using namespace smt;
+using namespace std;
+
+namespace pono {
+
+template <class Prover_T>
+CegarOpsUf<Prover_T>::CegarOpsUf(const Property & p,
+                                 const TransitionSystem & ts,
+                                 const SmtSolver & solver,
+                                 PonoOptions opt)
+  : super(p, create_fresh_ts(ts.is_functional(), solver), solver, opt),
+    conc_ts_(ts, super::to_prover_solver_),
+    prover_ts_(super::prover_interface_ts()),
+    oa_(conc_ts_, prover_ts_),
+    cegopsuf_solver_(create_solver(solver->get_solver_enum())),
+    to_cegopsuf_solver_(cegopsuf_solver_),
+    from_cegopsuf_solver_(super::prover_interface_ts().solver()),
+    cegopsuf_ts_(cegopsuf_solver_),
+    cegopsuf_un_(cegopsuf_ts_)
+{
+  cegopsuf_solver_->set_opt("produce-unsat-cores", "true");
+}
+
+template <class Prover_T>
+void CegarOpsUf<Prover_T>::set_ops_to_abstract(
+    const UnorderedOpSet & ops_to_abstract)
+{
+  oa_.set_ops_to_abstract(ops_to_abstract);
+}
+
+template <class Prover_T>
+ProverResult CegarOpsUf<Prover_T>::check_until(int k)
+{
+  initialize();
+
+  ProverResult res = ProverResult::FALSE;
+  while (res == ProverResult::FALSE) {
+    // need to call parent's check_until in case it
+    // is another cegar loop rather than an engine
+    res = super::check_until(k);
+
+    if (res == ProverResult::FALSE) {
+      if (!cegar_refine()) {
+        return ProverResult::FALSE;
+      }
+    }
+  }
+
+  if (res == ProverResult::TRUE && super::invar_) {
+    // update the invariant
+
+    // TODO currently getting some errors in the concretization phase.
+    // Commenting for now.
+
+    //super::invar_ = oa_.concrete(super::invar_);
+    super::invar_ = NULL;
+  }
+
+  return res;
+}
+
+template <class Prover_T>
+void CegarOpsUf<Prover_T>::initialize()
+{
+  if (super::initialized_) {
+    return;
+  }
+
+  // specify which cegar_abstract in case
+  // we're inheriting from another cegar algorithm
+  CegarOpsUf::cegar_abstract();
+  super::initialize();
+
+  // update local version of ts over fresh solver
+  cegopsuf_ts_ = TransitionSystem(prover_ts_, to_cegopsuf_solver_);
+  // update cache
```

Comment:
Just FYI this doesn't handle UFs in the formula (because they're not an input or state variable). This is a problem elsewhere in the code that we need to address later so I don't think it needs to be fixed now.